### PR TITLE
Fixed issue when trying to check if a multidimensional array is dirty…

### PR DESCRIPTION
**Pull request by speedplli**
Found a bug in validating if the attribute is dirty in multidimensional-array it will always return true even if it's not changed
 
| Q             | A
| ------------- | ---
| Is it bugfix?    | ✔️
| New feature?  | ❌
| Breaks BC?    | ❌
| Fixed issues  | NONE

### CODE REVIEW

File: framework/db/BaseActiveRecord.php [hunk 1]
```diff
@@ -1760,4 +1760,20 @@ private function setRelationDependencies($name, $relation, $viaRelationName = nu
...
+     */
+    private function isAttributeDirty($attribute, $value)
+    {
+        $old_attribute = $this->oldAttributes[$attribute];
+        if (is_array($value) && is_array($this->oldAttributes[$attribute])) {
+            ArrayHelper::sortMultidimensionalArray($value);
+            ArrayHelper::sortMultidimensionalArray($old_attribute);
+        }
+
+        return $value !== $old_attribute;
```
Comment (WinterSilence) [inline framework/db/BaseActiveRecord.php:1777]: `$value != $old_attribute;` give same result without sorting

File: framework/helpers/BaseArrayHelper.php [hunk 2]
```diff
@@ -999,4 +999,19 @@ public static function filter($array, $filters)
...
     }
+
+    /**
+     * Sorts multidimensional array
+     * @param $array
+     * @return void
+     */
+    public static function sortMultidimensionalArray(&$array)
+    {
+        ksort($array);
```
Comment (WinterSilence) [inline framework/helpers/BaseArrayHelper.php:1010]: sorry, but it's work only for assoc arrays, universal solution:
~~~php
if (static::isIndexed($array)) {
   sort($array);
} else {
   ksort($array);
}
~~~

Comment (bizley) [reply framework/helpers/BaseArrayHelper.php:1010]: Why? ksort works for int-indexed arrays.

Comment (WinterSilence) [reply framework/helpers/BaseArrayHelper.php:1010]: @bizley index array - sort by values, assoc array - sort by keys:
~~~php
$oldValue = ['foo' => ['bar' => 1, 'baz' => 2]];
$newValue = ['foo' => ['baz' => 2, 'bar' => 1]];

$oldValue = [['bar' => 1], ['baz' => 2]];
$newValue = [['baz' => 2], ['bar' => 1]];
~~~

Comment (bizley) [reply framework/helpers/BaseArrayHelper.php:1010]: Yes, I know. @speedplli please elaborate on this step here.

Comment (WinterSilence) [reply framework/helpers/BaseArrayHelper.php:1010]: and add tests

Comment (speedplli) [reply framework/helpers/BaseArrayHelper.php:1010]: @bizley I will.
@WinterSilence Thanks I will implement.

Comment (speedplli) [reply framework/helpers/BaseArrayHelper.php]: @bizley @WinterSilence 
Done Please check. 
Tests were also added.

Thanks

File: tests/framework/helpers/ArrayHelperTest.php [hunk 3]
```diff
@@ -1436,6 +1436,28 @@ public function testArrayAccessWithMagicProperty()
         $this->assertEquals(42, ArrayHelper::getValue($model, 'magic'));
         $this->assertEquals('ta-da', ArrayHelper::getValue($model, 'moreMagic'));
     }
+
+    public function testSortMultidimensionalArray()
```
Comment (WinterSilence) [inline tests/framework/helpers/ArrayHelperTest.php:1440]: please, use data provider, see [Example 2.6 Using a data provider with named datasets](https://phpunit.readthedocs.io/en/7.5/writing-tests-for-phpunit.html#writing-tests-for-phpunit-data-providers-examples-datatest1-php) and add complex data to tests:
~~~php
['c' => ['bar' => 1, 'baz' => 1], 'a' => ['baz' => 2, 'bar' => 1], 'b' => '...'];

[['attr', 'attr2'], 'equal', 'option' => 'value', 'option2' => 'value2'];
[['attr2', 'attr'], 'equal', 'option2' => 'value2', 'option' => 'value'];
~~~

File: framework/helpers/BaseArrayHelper.php [hunk 4]
```diff
@@ -999,4 +999,24 @@ public static function filter($array, $filters)
...
+        } else {
+            ksort($array);
+        }
+
+        foreach ($array as &$item) {
+            if (is_array($item)) {
+                self::sortMultidimensionalArray($item);
+            }
+        }
+    }
```
Comment (WinterSilence) [inline framework/helpers/BaseArrayHelper.php:1021]: `return $array;` to using as callback for `array_*` functions

File: framework/helpers/BaseArrayHelper.php [hunk 5]
```diff
@@ -999,4 +999,24 @@ public static function filter($array, $filters)
...
     }
+
+    /**
+     * Sorts multidimensional array
+     * @param $array
+     * @return void
+     */
+    public static function sortMultidimensionalArray(&$array)
+    {
+        if (static::isIndexed($array)) {
```
Comment (WinterSilence) [inline framework/helpers/BaseArrayHelper.php:1010]: you must sort index array after sorting values

**Comment by WinterSilence** [comment]
combined impementation:
~~~php
/**
 * Sorts array recursively.
 *
 * @param array $array An array passing by reference.
 * @param callable|null $sorter The array sorter. If ommit, sort index array by values, sort assoc array by keys.
 * @return array
 */
public static function recursiveSort(array &$array, $sorter = null)
{
    foreach ($array as &$value) {
        if (is_array($value)) {
            self::recursiveSort($value, $sorter);
        }
    }
    unset($value);

    if ($sorter === null) {
        if (static::isIndexed($array)) {
            sort($array);
        } else {
            ksort($array);
        }
    } else {
        call_user_func_array($sorter, [&$array]);
    }

    return $array;
}
~~~

**Comment by WinterSilence** [comment]
if arrays contain float values it's can be fail comparation

File: tests/framework/helpers/ArrayHelperTest.php [hunk 6]
```diff
@@ -1436,6 +1436,28 @@ public function testArrayAccessWithMagicProperty()
...
+        ArrayHelper::sortMultidimensionalArray($empty_array);
+        $this->assertEquals([], $empty_array);
+
+        // assoc array
+        $assoc_array = ['foo' => ['bar' => 1, 'baz' => 2]];
+        ArrayHelper::sortMultidimensionalArray($assoc_array);
+        $this->assertEquals(['foo' => ['baz' => 2, 'bar' => 1]], $assoc_array);
+
+        if (PHP_VERSION_ID < 70000) {
+            $this->markTestSkipped('Can not be tested on PHP < 7.0');
```
[the remark this answers is not in the document]
Comment (speedplli) [reply tests/framework/helpers/ArrayHelperTest.php:1453]: When I run the test it fails on PHP 5.X versions
I will implement it in a new way and add extra tests

[the remark this answers is not in the document]
Comment (WinterSilence) [reply tests/framework/helpers/ArrayHelperTest.php:1453]: @speedplli no, your expected arrays is invalid, see
https://3v4l.org/h1YS5#v5.4.45
https://3v4l.org/h1YS5#v7.1.33

File: tests/framework/helpers/ArrayHelperTest.php (hunk 3 above)
Comment (speedplli) [reply tests/framework/helpers/ArrayHelperTest.php]: @WinterSilence Please check the modifications for the new test cases

File: tests/framework/helpers/ArrayHelperTest.php [hunk 7]
```diff
@@ -1437,26 +1437,70 @@ public function testArrayAccessWithMagicProperty()
...
+            [
+                [1, 2, 3, 4],
+                [4, 1, 3, 2]
+            ],
+            //Normal associative array
+            [
+                ['a' => 1, 'b' => 2, 'c' => 3, 'd' => 4],
+                ['b' => 2, 'a' => 1, 'd' => 4, 'c' => 3],
+            ],
+            //Normal index array
```
Comment (WinterSilence) [inline tests/framework/helpers/ArrayHelperTest.php:1468]: duplicate

Comment (WinterSilence) [reply tests/framework/helpers/ArrayHelperTest.php:1468]: you can use something like [yii2-faker](https://github.com/yiisoft/yii2-faker/) to generate random data. add mixed arrays in provider, you can add you db data

File: framework/helpers/BaseArrayHelper.php [hunk 8]
```diff
@@ -7,9 +7,9 @@
 
 namespace yii\helpers;
 
-use ArrayAccess;
-use Traversable;
 use Yii;
+use Traversable;
+use ArrayAccess;
```
Comment (bizley) [inline framework/helpers/BaseArrayHelper.php:12]:
```suggestion
use ArrayAccess;
use Traversable;
```

**Review by bizley — APPROVED** [review]

**Review by Arhell — APPROVED** [review]

**Comment by samdark** [comment]
Would you please add a line for CHANGELOG?

**Comment by speedplli** [comment]
> Would you please add a line for CHANGELOG?

Yes, Added.

**Comment by bizley** [comment]
Thank you.

**Pull request merged**